### PR TITLE
Reintroduce Shell API to launch applications

### PR DESCRIPTION
**Pull request by andrunko**
https://phabricator.endlessm.com/T26726

### CODE REVIEW

File: js/ui/shellDBus.js [hunk 1]
```diff
@@ -606,3 +611,80 @@ var HackableAppsManager = class {
...
+<method name="LaunchViaDBusCall"> \
+    <arg type="s" direction="in" name="name" /> \
+    <arg type="s" direction="in" name="busName" /> \
+    <arg type="s" direction="in" name="objectPath" /> \
+    <arg type="s" direction="in" name="interfaceName" /> \
+    <arg type="s" direction="in" name="methodName" /> \
+    <arg type="v" direction="in" name="args" /> \
+</method> \
+</interface> \
+</node>';
```
Comment (GeorgesStavracas) [inline js/ui/shellDBus.js]: Please move this to under `data/dbus-interfaces/org.gnome.Shell.AppLauncher.xml` and load it with `loadInterfaceXML('org.gnome.Shell.AppLauncher')`

File: js/ui/shellDBus.js [hunk 2]
```diff
@@ -606,3 +611,80 @@ var HackableAppsManager = class {
...
+
+function activationContextForAppName(appName, appSys) {
```
Comment (GeorgesStavracas) [inline js/ui/shellDBus.js]: You can move this function to inside `AppLauncher` and remove the second argument

File: js/ui/appActivation.js [hunk 3]
```diff
@@ -85,6 +85,44 @@ var AppActivationContext = class {
...
+
+            let invoke = (() => {
+                proxy.call(method, args, Gio.DBusCallFlags.NONE, -1, null, function(source, result) {
+                    try {
+                        done(null, proxy.call_finish(result));
+                    } catch (error) {
+                        done(error, null);
+                    }
+                });
+            });
```
Comment (GeorgesStavracas) [inline js/ui/appActivation.js]: You can unwrap this to simply

```javascript
	                proxy.call(method, args, Gio.DBusCallFlags.NONE, -1, null,(source, result) => {
	                    try {
	                        done(null, proxy.call_finish(result));
	                    } catch (error) {
	                        done(error, null);
	                    }
	                });
```

and move it to below `this.showSplash();` (which is where this method is invoked)

File: js/ui/appActivation.js [hunk 4]
```diff
@@ -85,6 +85,44 @@ var AppActivationContext = class {
...
+                g_connection: bus,
+                g_default_timeout: GLib.MAXINT32,
+                g_flags: Gio.DBusProxyFlags.NONE,
+                g_interface_name: interfaceName,
+                g_name: busName,
+                g_object_path: path
+            });
+
+            let invoke = (() => {
+                proxy.call(method, args, Gio.DBusCallFlags.NONE, -1, null, function(source, result) {
```
Comment (GeorgesStavracas) [inline js/ui/appActivation.js]: `function(source, result)` → `(source, result) =>`

File: js/misc/eos-launch.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    // trim .desktop part, if present
+    let idx = appName.indexOf('.desktop');
+    if (idx != -1)
+        appName = appName.substring(0, idx);
+
+    let appNames = [appName];
+
+    let languageNames = GLib.get_language_names();
+    let variants = GLib.get_locale_variants(languageNames[0]);
+    variants.filter(function(variant) {
```
Comment (GeorgesStavracas) [inline js/misc/eos-launch.js]: Use arrow function (`(variant) => {`)

File: js/misc/eos-launch.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        // discard variants with an encoding
+        return (variant.indexOf('.') == -1)
+    }).forEach(function(variant) {
```
Comment (GeorgesStavracas) [inline js/misc/eos-launch.js]: Use arrow function (`(variant) => {`)

File: js/ui/appActivation.js [hunk 7]
```diff
@@ -85,6 +85,44 @@ var AppActivationContext = class {
...
+            }
+
+            let proxy = new Gio.DBusProxy({
+                g_bus_type: Gio.BusType.SESSION,
+                g_connection: bus,
+                g_default_timeout: GLib.MAXINT32,
+                g_flags: Gio.DBusProxyFlags.NONE,
+                g_interface_name: interfaceName,
+                g_name: busName,
+                g_object_path: path
```
Comment (GeorgesStavracas) [inline js/ui/appActivation.js]: Nitpick: add trailing comma

**Comment by andrunko** [comment]
@GeorgesStavracas thanks for the review, I updated the branch, please check again.

**Pull request merged**